### DISCUSSION TITLE
How to open the DatePicker/TimePicker dialog in code?

### DISCUSSION THREAD

**Discussion by joaowatchful** [dotnet/maui#7410] [Q&A]
I'm trying to create a custom DateTimerPicker control where you select both date and time to update an entry. Both the DatePicker and TimePicker are not visible and I'm trying to get both controls' dialogs to open but I can't seem to find a way. I tried with .Focus() but it doesn't seem to work.

**Comment by Ghostbird** [comment]
I'm having the same problem. I ported some Xamarin code that used `Focus()`, but it seems there's no way to programmatically (either C# or Xaml) open these pickers.
Also the exact same use case. A DateTimePicker control, because Xamarin/MAUI never had one.

**Reply by Ghostbird** [reply]
The issue is tracked [here](../issues/8946), the old `Focus()` behaviour had some (in my opinion logical) side-effects. Someone complained [about a side-effect](../issues/8130) and it was “[fixed](../pull/8913)”. I put fixed in quotes, since instead of fixing the behaviour, the ability to focus a date/time picker was essentially entirely removed.

**Reply by AndyG74** [reply]
i have the same problem.

**Reply by needledrag** [reply]
Same, same..

**Reply by beglaryan-alen** [reply]
same....

**Comment by Sergtek** [comment]
I have found a way to open the date picker, Android's native View class (Android.Views) have a method called PerformClick which allows opening the date picker without pressing the DatePicker and TimePicker, I leave you an example of how I did it in case it can help someone in the same situation.

I have made a small example in which I have a DatePicker, when the user clicks on it the date picker appears, when we select a date from the picker and click "OK" the DateSelected event is fired and this is where the magic is done to access the TimePicker Handler and thus be able to access the native "PerformClick" method to open the TimePicker time selector without having to click on it:

**XAML:**
```
        <StackLayout
            HorizontalOptions="CenterAndExpand"
            VerticalOptions="CenterAndExpand">
            <DatePicker 
                x:Name="datePicker"
                DateSelected="datePicker_DateSelected"
                BackgroundColor="White"
                TextColor="Black"
                FontSize="20"
                WidthRequest="200"
                HeightRequest="200">
            </DatePicker>

            <TimePicker 
                x:Name="timePicker"
                IsVisible="False"/>
        </StackLayout>
```

**Code-behind:**
```
    private void datePicker_DateSelected(object sender, DateChangedEventArgs e)
    {
#if ANDROID
        var handler = timePicker.Handler as ITimePickerHandler;
        handler.PlatformView.PerformClick();
#endif
    }
```

**Image:**
![fixFocusIssue](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif)

**Reply by LtFinTutuola** [reply]
You are a fucking **GENIUS**!!!
This issue took me a plenty of time, but your approach solve the problem on the root! GG

**Reply by jedipi** [reply]
OMG!!! This resolved my issue on Android.
Do you have a solution for IOS as well??

**Reply by h1nson** [reply]
You saved my day!

**Reply by Kuurse** [reply]
@jedipi Did you find a way to do this on iOS?

**Reply by nunosantosgithub** [reply]
On iOS you can just call `timePicker.Focus();`

**Reply by mfranc28** [reply]
but PerformClick don't exist.....how it works?

**Comment by mahop-net** [comment]
This is still an issue in .Net 8. Will there be a fix?

**Reply by Ghostbird** [reply]
I'm pretty sure there won't be one. The current situation _is_ the fixed behaviour https://github.com/dotnet/maui/pull/8913. That PR suggests it was never intended that you could open a date/time picker in any way, except by direct user interaction.

**Reply by Ghostbird** [reply]
However, https://github.com/dotnet/maui/issues/8945 suggests that a new API may be added to add a way to open these types of controls from code. Note that this won't fix the focus behaviour. Depending on how hard you need it, you could make a PR yourself. The main difficulty I think with making such a PR is that you'll have to cover all supported platforms.

**Reply by mahop-net** [reply]
@Ghostbird - Thanks for your answers!

Meanwhile I'm using Telerik Controls... I ported a bigger project form Xamarin to Maui and was trying to remove the dependency to Telerik Controls. But I gave up and use them again. The Controls provided by Maui have sometimes less funtionality compared to Xamarin... The good thing was, that there are no Layout Bugs in Maui. I was able to remove all the layout tweaks used in the Xamarin project...

(And for new projects I'm using flutter or WebView with pure JS. After almost 30 years of "Microsoft" development (VB 3 - VB6 and C# .net 1.0 until now) I'm finally switching more and more...)

**Comment by YakushenkoES** [comment]
Is there any solution for Windows?

**Reply by formerlymisterhenson** [reply]
I am also trying to make my own DateTimePicker work with Windows.

DatePicker is somewhat working:
```
            MainThread.BeginInvokeOnMainThread(() =>
            {
#if WINDOWS
                var handler = datePicker.Handler as IDatePickerHandler;
                handler.PlatformView.IsCalendarOpen = true;
#elif IOS
                datePicker.Focus();
#endif
            });
```
For the TimePicker I spent much time, but there is no similiar way. Invoking OnTapped with Reflection didn't work either. Needless to say this ugly. There's the claim that doing this with Focus() would cause issues with screen readers (accessibility), but that explanation seems somewhat like an excuse. If a UI framework can't manage to get basic features like a datetime picker shipped, then accessibility seems to me a bit off the scope.
